### DISCUSSION TITLE
Berry - convert string to a list

### DISCUSSION THREAD

**Discussion by xostia** [arendst/Tasmota#20751] [Support request] [answered]
Hi,

Anybody knows how to convert a string to a list in berry?. I am using this code (bottom), but it should be a easier way (list to string is simply "tostring()"). I also tried to convert with list (), same style than number() ... but it does not work (even using the same string generated by tostring()).

def stringToList(list_str)
    import string
    var result = []
    var temp = string.split(string.split(string.split(list_str, -1)[0], '[')[1], ',')
    for i:temp
        result.push(number(i))
    end
    return result
end

[Edit]

The conversions that I get with list() simply take the list as one single string element

    print (stringToList(readingsMap[IR_RAW_LABEL]["value"]))
  > I get:  [65535, 65535, 65535, 65535, 65535, 65535]

    print (list(readingsMap[IR_RAW_LABEL]["value"]))
 > I get ['[65535, 65535, 65535, 65535, 65535, 65535]']

Thank you

**Comment by sfromis** [comment]
Well, the conversion depends on exactly how the string can be, but your code looks like it can be JSON-like. Hence:
```
var list_str='[1,2,3,4]'
import json
var result = json.load(list_str)
```
To create a suitable string from more than a simple list of numbers, I can recommend using `json.dump` instead of `tostring()`.

You added more stuff while I was typing, but how `readingsMap` was created could also be a factor in how to process it.

**Reply by xostia** [reply]
Thank you sfromis...

We created this function to convert because the readingsMap is a map and this field is filled with "" (string) and I did not find other way to create a list from it.

Copying this field to other string, it appeared with "" and in MQTT message I wanted it as a list (HA interpretes ok the [1,2,3,4] format].

It is not the target of the thread but... I tried to use "mqtt.publish" (tried also tasmota.publish) and in the console seem ok but HA does not receive the messages... do yo know why?, any limitation about where I can execute mqtt.publish (I tested in autoexec.be after mqtt connection, inside a EVERY_MINUTE in a driver, inside the JSON update function... nothing).

I want to add some fields calculated in Berry and taken from the MQTT message (calculations based on sensor values in C). Is there any way to detect "just before the MQTT messaging" to perform these simple calculations and inject the fields in the MQTT message?

Thank you again.

**Reply by sfromis** [reply]
If the message is verifiably published via MQTT, and not picked up by HA, it could be because it was not published to a topic HA subscribed to. Using `mqtt.publish` is preferred, as `tasmota.publish` is explicitly marked as "deprecated".

In general, I can't do anything with words like "seems ok", and "does not", as lots and lots of issues are about spotting the detail which was not "quite right".

Even if you did not tell what messages you're talking about (specifics are much better than abstract generalities), I'm assuming that you mean the sensor messages published at TelePeriod. What you can do in Berry is to have a class instance registered as a driver, and providing JSON for an additional driver object in the sensors payload. This does not allow you to read the other message from *the same* payload, but you could have a separate function to poll the values, and thus have a basis for your additional sensor data. While the "original sensor" values might not be 100% the same as what you picked up a second earlier, it would be very close in the large majority of cases around temperatures.

**Reply by xostia** [reply]
OK, sfromis... I am sorry, I will try to be more specific

This is a regular message from my tasmota, received by HA without problem
11:16:28.829 MQT: tele/FDSM_7844B0/SENSOR = {"Time":"2024-02-19T11:16:28+00:00","ANALOG":{"A1":2832},"AHT2X":etc. etc. et.

This is the message sent "by hand" with mqtt.publish (right now is a fix string). In HA, Magnitudes and CalibrationCoefs never appear.

11:16:28.823 MQT: tele/FDSM_7844B0/SENSOR = {"Magnitudes":22,"CalibrationCoefs":[1,2,3]}

Thank you for your patience.

**Reply by sfromis** [reply]
Do not expect HA to pick up on your faked sensors payloads, without manual config. You should instead use the Berry features to have it included in the real ones. The automatic detection in HA relies on an extra discovery payload after boot, including all sensors.

**Reply by xostia** [reply]
I understand, sfromis...

This "fake sensor" is a calculation on data provided by TSL2591... if I "inject" the data in the Tasmota MQTT message, these calculations will arrive in the next MQTT message (the "fake sensor" calculation will be outdate, based on the "old data" in the previous message).

My idea was to send a separate MQTT message with only these calculated data using mqtt.publish.

For htis, It would be interesting to have an event when the MQTT message is ready to go... but not jet send.

I can live with that. Thank you.

**Reply by sfromis** [reply]
The "old data" would not have to be from the previous message, more like from one second ago.

By "fake" I was not talking about the content of your calculation, but about the action of doing the separate publish to a topic managed by Tasmota.

**Reply by xostia** [reply]
It is true... in a regular sensor, not this case.

The light sensor needs to work in max gain configuration and long time exposition (600ms), then I configured the code of TSL2591 to only perform the readings in the last 10 seconds of the teleperiod (and teleperiod 60 seconds).

If I inject the results of the calculation, they will appear "one minute later" because they are based on the sensor data "one minute before" (in Berry I check the updates of MQTT with callback... just after the MQTT message is sent).

This is not part of the discussion, but I am considering to use this code for the TSL2591 (and not the Arduino library):

https://github.com/gmarti/AsyncTSL2591

The integration time will not have impact in Tasmota because is totally async.

What I am doing is a photometer for night light measurement (light contamination). We made a complete design of the device with cheap components. If you have interest, I can tell you more.

Thank you

**Reply by sfromis** [reply]
Well, when you are deliberately breaking the driver to not be able to work as usual, it is up to you to make it work.

You could do something like starting the sampling 11 seconds before TelePeriod, and then have Berry code to pick up the value 1 second before TelePeriod, do your additional calculations, and then have your `json_append` callback supply the extra object for the sensors payload.

**Reply by xostia** [reply, marked as answer]
Yes... that is what I am thinking... I was checking and the measurements start second 50/51, then it should be easy to calculate the moment to launch json_append.

Thank you a lot.
